torchao/prototype/intx/bitpacking.py

```diff
@@ -0,0 +1,200 @@
+    4: (0x0f,),
+    5: (0x0f, 0x10),
+    6: (0x0f, 0x30),
+    7: (0x0f, 0x30, 0x40),
+}
+
+unpack_mask = {
+    1: (0x01,0x02,0x04,0x08, 0x10,0x20,0x40,0x80),
+    2: (0x03,0x0c,0x30,0xc0),
+    4: (0x0f,0xf0),
+}
+
+# size of each shard
+numbits = {
+    1: (1,),
+    2: (2,),
+    3: (2, 1),
+    4: (4,),
+    5: (4, 1),
+    6: (4, 2),
+    7: (4, 2, 1),
+}
+
+# shift amount for each shard
+shifts = {
+    1: (0,),
+    2: (0,),
+    3: (0, 2),
+    4: (0,),
+    5: (0, 4),
+    6: (0, 4),
+    7: (0, 4, 6),
+}
+
+# for shifting groups left but right if shift is negative
+def abs_lsh(data, shift):
+    if shift == 0:
+        return data
+    elif shift < 0:
+        return data >> -shift
+    else:
+        return data << shift
+
+
+# inverse of abs_lsh for unpacking
+def abs_rsh(data, shift):
+    if shift == 0:
+        return data
+    elif shift < 0:
+        return data << -shift
+    else:
+        return data >> shift
+
+
+def pack_cpu(data: torch.Tensor,
+         elem_size: int,
+         dim: Optional[int] = -1) -> List[torch.Tensor]:
+    """
+    Inputs:
+    data: a tensor of sub byte elements in uint8 
+    dim: the dimension to pack along
+    Returns: a list of packed shards
+    
+    given an array such as [0x30,0x29,0x17,0x5,0x20,0x16,0x9,0x2] which are 8 uint6 elements
+    first seperate into two shards: the upper 2 bits and the lower 4 bits by using a mask (0x30 and 0x0f respectively)
+    2 bit shards: [0b00110000, 0b00100000, 0b00010000, 0b00000000, 0b00100000, 0b00010000, 0b00000000, 0b00000000]
+                   |------ group 1 ------| |------ group 2 ------| |------ group 3 ------| |------ group 4 ------|
+    4 bit shards: [0b00000000, 0b00001001, 0b00000111, 0b00000101, 0b00000000, 0b00000110, 0b00001001, 0b00000010]
+                   |------------------ group 1 ------------------| |------------------ group 2 ------------------|
+    Then pack each of these shards by shifting each group of 4 uint2s or 2 uint4s to a different position within the 8bit container
+    2bit shard group1 >> 4,  group2 >> 2, group3 >> 0, group4 << 2
+    [0b00000011, 0b00000001, 0b00000100, 0b00000000, 0b00100000, 0b00010000, 0b00000000, 0b00000000]
+    |------ group 1 ------| |------ group 2 ------| |------ group 3 ------| |------ group 4 ------|
+    4 bit shard group1 << 0, group2 << 4
+    [0b00000000, 0b00001001, 0b00000111, 0b00000101, 0b00000000, 0b01100000, 0b10010000, 0b00100000]
+    |------------------ group 1 ------------------| |------------------ group 2 ------------------|
+    finally bitwise or the groups together
+    2 bit shard: [0b00100111, 00010001]
+    4 bit shard: [0b00000000, 0b01101001, 0b10010111, 0b00100101]
+    so we went from 8 elements to 6
+    in general this means we go from 8 * n to elem_size * n
+    """
+    torch._assert(data.shape[dim] % 8 == 0, f"pack dimension size ({data.shape[dim]}) is not divisble by scale")
+    if data.dtype != torch.uint8:
+        data = data.to(torch.uint8)
```

Comment:
should we just error out?

Reply:
I don't even have this check in the GPU kernel but yea I can change to error out in both